packages/runtime/src/useSSRComputation_Client.ts

```diff
@@ -0,0 +1,120 @@
+import { useCallback, useEffect, useMemo, useRef, useState } from "react";
+import {
+  calculateCacheKey,
+  Dependency, isObservable,
+  isPromise, Observable,
+  Options,
+  parseDependencies,
+  SSRComputationFunction, Subscription,
+} from "./utils";
+import { wrapErrorHandler } from "./errorHandler";
+import { getSSRCache } from "./ssrCache";
+import { runOnSubscriptionsResumed } from "./subscriptions";
+import { ClientComputationFunction } from "./utils";
+
+const useSSRComputation_Client = <TResult>(
+  importFn: ClientComputationFunction<TResult>,
+  dependencies: Dependency[],
+  options: Options,
+  relativePathToCwd: string
+): TResult | null => {
+  const [, forceUpdate] = useState(0);
+  const cache = getSSRCache();
+  const parsedDependencies = parseDependencies(dependencies);
+  const skip = !!options.skip;
+  const subscriptionRef = useRef<Subscription>()
+  const fnRef = useRef<SSRComputationFunction<TResult>>()
+  const isMountedRef = useRef(true);
+  const currentCacheKeyRef = useRef('');
+  const currentResultRef = useRef<TResult | null>(null);
+
+  // relativePathToCwd is used to make sure that the cache key is unique for each module
+  // and it's not affected by the file that calls it
+  const cacheKey = calculateCacheKey(relativePathToCwd, parsedDependencies);
+  currentCacheKeyRef.current = cacheKey;
+  const cachedResult = cache[cacheKey]?.result as TResult | undefined;
+  const isSubscription = cache[cacheKey]?.isSubscription;
+  const isCacheHit = cacheKey in cache;
+
+  useEffect(() => {
+    return () => {
+      isMountedRef.current = false;
+      subscriptionRef.current?.unsubscribe();
+      subscriptionRef.current = undefined;
+    }
+  }, [cacheKey]);
+
+  const isDisposed = useCallback(() => {
+    return cacheKey !== currentCacheKeyRef.current || !isMountedRef.current;
+  }, [cacheKey]);
+
+  const handleResult = useCallback((returnedResult: TResult | Promise<TResult> | Observable<TResult>) => {
+    if (isDisposed()) return;
+
+    const updateResult = (newResult: TResult) => {
+      if (isDisposed()) return;
+      if (currentResultRef.current === newResult) return;
+      cache[cacheKey] = {
+        result: newResult,
+        isSubscription: cache[cacheKey]?.isSubscription || isObservable(returnedResult),
+      };
+      forceUpdate(prevState => prevState + 1);
+    }
+
+    if (isPromise(returnedResult)) {
+      returnedResult.then(updateResult);
+    } else if (isObservable(returnedResult)) {
+      subscriptionRef.current = returnedResult.subscribe({ next: updateResult });
+    } else {
+      updateResult(returnedResult);
+    }
+  }, [forceUpdate, isDisposed, cacheKey]);
+
+  const loadAndRun = useCallback(() => {
+    if (!isMountedRef.current) return;
+    importFn().then(module => {
+      if (!isMountedRef.current) return;
+      fnRef.current = module.default;
+      handleResult(module.default(...parsedDependencies));
```

Comment:
If I get the code right, [this performance issue](https://github.com/Popmenu/use-ssr-computation.macro/pull/70#discussion_r1328716726
) is still unresolved, right?

Why can't we export the `subscribe` function separately?
Why do we need to call the `SSRComputationFunction` after we load the module? The function can be CPU heavy, if we have the SSR cache, why can't we use it instead of calling the function?

Reply:
Okay, I misread the code. It's not a performance issue.

Still, I find this API hard to document: when the `default computation` function is called client-side, the hook may ignore the `TResult.current` value, which may confuse people

This confusion may be resolved by enforcing a single correct return value to `default computation` functions. E.G.:
```
export type SSRComputationFunction<TResult> = {
  //  `compute` will be called in the following cases
  // * when `useSSRComputation` hook is invoked server-side
  // * client-side if the hook wasn't called server-side
  // * client-side when `dependencies` change
  compute: (dependencies: Dependencies) => TResult | null;
  // Optional `subscribe` is never called server-side. Client-side, it will be called in the following cases:
  // * when `fetchSubscriptions` is called
  // * when `dependencies` change (right after `compute` call)
  subscribe?: (getCurrentResult: () => TResult, dependencies: Dependencies, next: (result: TResult)=>void ) => Subscription;
}
```

This way, we will have a short and well-documented API.